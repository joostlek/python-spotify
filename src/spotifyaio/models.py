"""Models for Spotify."""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin


class DeviceType(StrEnum):
    """Device type."""

    COMPUTER = "Computer"
    SMARTPHONE = "Smartphone"
    SPEAKER = "Speaker"


@dataclass
class Device(DataClassORJSONMixin):
    """Device model."""

    device_id: str = field(metadata=field_options(alias="id"))
    is_active: bool
    is_private_session: bool
    is_restricted: bool
    name: str
    device_type: DeviceType = field(metadata=field_options(alias="type"))
    volume_percent: int
    supports_volume: bool = True


@dataclass
class Devices(DataClassORJSONMixin):
    """Devices model."""

    devices: list[Device]


class RepeatMode(StrEnum):
    """Repeat mode."""

    OFF = "off"
    TRACK = "track"
    CONTEXT = "context"


class ContextType(StrEnum):
    """Context type."""

    ALBUM = "album"
    ARTIST = "artist"
    PLAYLIST = "playlist"
    COLLECTION = "collection"
    SHOW = "show"


@dataclass
class Context(DataClassORJSONMixin):
    """Context model."""

    external_urls: dict[str, str]
    href: str
    context_type: ContextType = field(metadata=field_options(alias="type"))
    uri: str


class AlbumType(StrEnum):
    """Album type."""

    ALBUM = "album"
    SINGLE = "single"
    COMPILATION = "compilation"


@dataclass
class Image(DataClassORJSONMixin):
    """Image model."""

    height: int | None
    width: int | None
    url: str


class ReleaseDatePrecision(StrEnum):
    """Release date precision."""

    YEAR = "year"
    MONTH = "month"
    DAY = "day"


@dataclass
class SimplifiedArtist(DataClassORJSONMixin):
    """Simplified artist model."""

    artist_id: str = field(metadata=field_options(alias="id"))
    name: str
    uri: str


@dataclass
class Album(DataClassORJSONMixin):
    """Album model."""

    album_type: AlbumType
    total_tracks: int
    album_id: str = field(metadata=field_options(alias="id"))
    images: list[Image]
    name: str
    release_date: str
    release_date_precision: ReleaseDatePrecision
    uri: str
    artists: list[SimplifiedArtist]


@dataclass
class Artist(DataClassORJSONMixin):
    """Artist model."""

    artist_id: str = field(metadata=field_options(alias="id"))
    name: str
    uri: str


@dataclass
class Track(DataClassORJSONMixin):
    """Track model."""

    track_id: str = field(metadata=field_options(alias="id"))
    album: Album
    artists: list[Artist]
    disc_number: int
    duration_ms: int
    explicit: bool
    external_ids: dict[str, str]
    external_urls: dict[str, str]
    href: str
    is_local: bool
    track_number: int
    object_type: str = field(metadata=field_options(alias="type"))
    uri: str


@dataclass
class CurrentPlaying(DataClassORJSONMixin):
    """Current playing model."""

    context: Context | None
    progress_ms: int | None
    is_playing: bool
    item: Track | None
    currently_playing_type: str | None


@dataclass
class PlaybackState(CurrentPlaying):
    """Playback state model."""

    device: Device
    shuffle: bool = field(metadata=field_options(alias="shuffle_state"))
    repeat_mode: RepeatMode = field(metadata=field_options(alias="repeat_state"))


class PlaylistOwnerType(StrEnum):
    """Playlist owner type."""

    USER = "user"


@dataclass
class PlaylistOwner(DataClassORJSONMixin):
    """Playlist owner model."""

    display_name: str
    external_urls: dict[str, str]
    href: str
    owner_id: str = field(metadata=field_options(alias="id"))
    object_type: str = field(metadata=field_options(alias="type"))
    uri: str


@dataclass
class BasePlaylist(DataClassORJSONMixin):
    """Base playlist model."""

    collaborative: bool
    description: str | None
    external_urls: dict[str, str]
    playlist_id: str = field(metadata=field_options(alias="id"))
    images: list[Image]
    name: str
    owner: PlaylistOwner
    public: bool | None
    object_type: str = field(metadata=field_options(alias="type"))
    uri: str


@dataclass
class Playlist(BasePlaylist):
    """Playlist model."""


@dataclass
class PlaylistResponse(DataClassORJSONMixin):
    """Playlist response model."""

    href: str
    items: list[BasePlaylist]
    limit: int
    next_list: str | None = field(metadata=field_options(alias="next"))
    offset: int
    previous_list: str | None = field(metadata=field_options(alias="previous"))
    total: int


@dataclass
class FeaturedPlaylistResponse(DataClassORJSONMixin):
    """Featured playlist response model."""

    message: str
    playlists: PlaylistResponse


@dataclass
class CategoryPlaylistResponse(DataClassORJSONMixin):
    """Category playlist response model."""

    playlists: PlaylistResponse


class ProductType(StrEnum):
    """Product type."""

    PREMIUM = "premium"
    FREE = "free"


@dataclass
class BaseUserProfile(DataClassORJSONMixin):
    """Base user profile model."""

    display_name: str
    user_id: str = field(metadata=field_options(alias="id"))
    images: list[Image]
    object_type: str = field(metadata=field_options(alias="type"))
    uri: str


@dataclass
class UserProfile(BaseUserProfile):
    """User profile model."""

    email: str
    product: ProductType
