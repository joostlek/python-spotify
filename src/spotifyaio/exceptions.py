"""Asynchronous Python client for Spotify."""


class SpotifyError(Exception):
    """Generic exception."""


class SpotifyConnectionError(SpotifyError):
    """Spotify connection exception."""


class SpotifyAuthenticationFailedError(SpotifyError):
    """Spotify authentication failed exception."""


class SpotifyNotFoundError(SpotifyError):
    """Spotify not found exception."""


class SpotifyRateLimitError(SpotifyError):
    """Spotify rate limit exception."""
