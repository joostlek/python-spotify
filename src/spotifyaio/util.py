"""Utility functions for the SpotifyAIO package."""


def get_identifier(identifier: str) -> str:
    """Get the identifier from a Spotify URI."""
    return identifier.split(":")[-1]
