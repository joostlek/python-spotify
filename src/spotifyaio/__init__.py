"""Asynchronous Python client for Spotify."""

from .exceptions import (
    SpotifyAuthenticationFailedError,
    SpotifyConnectionError,
    SpotifyError,
    SpotifyNotFoundError,
    SpotifyRateLimitError,
)
from .models import (
    Album,
    AlbumType,
    Artist,
    BasePlaylist,
    BaseUserProfile,
    Context,
    ContextType,
    CurrentPlaying,
    Device,
    DeviceType,
    Episode,
    Image,
    Item,
    ItemType,
    PlaybackState,
    Playlist,
    PlaylistOwner,
    PlaylistOwnerType,
    ProductType,
    ReleaseDatePrecision,
    RepeatMode,
    Show,
    SimplifiedAlbum,
    SimplifiedArtist,
    SimplifiedTrack,
    Track,
    UserProfile,
)
from .spotify import SpotifyClient

__all__ = [
    "Album",
    "AlbumType",
    "Artist",
    "BasePlaylist",
    "BaseUserProfile",
    "Context",
    "ContextType",
    "CurrentPlaying",
    "Device",
    "DeviceType",
    "Episode",
    "Image",
    "Item",
    "ItemType",
    "PlaybackState",
    "Playlist",
    "PlaylistOwner",
    "PlaylistOwnerType",
    "ProductType",
    "ReleaseDatePrecision",
    "RepeatMode",
    "Show",
    "SimplifiedAlbum",
    "SimplifiedArtist",
    "SimplifiedTrack",
    "SpotifyAuthenticationFailedError",
    "SpotifyClient",
    "SpotifyConnectionError",
    "SpotifyError",
    "SpotifyNotFoundError",
    "SpotifyRateLimitError",
    "Track",
    "UserProfile",
]
