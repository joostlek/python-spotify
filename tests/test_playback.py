"""Asynchronous Python client for Spotify."""
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from aiohttp.hdrs import METH_GET, METH_POST, METH_PUT
from aioresponses import aioresponses
import pytest
from yarl import URL

from spotifyaio import RepeatMode

from . import load_fixture
from .const import HEADERS, SPOTIFY_URL

if TYPE_CHECKING:
    from spotifyaio import SpotifyClient
    from syrupy import SnapshotAssertion


@pytest.mark.parametrize(
    "playback_id",
    [
        1,
        2,
        3,
    ],
)
async def test_get_playback_state(
    responses: aioresponses,
    snapshot: SnapshotAssertion,
    playback_id: int,
    authenticated_client: SpotifyClient,
) -> None:
    """Test retrieving playback state."""
    responses.get(
        f"{SPOTIFY_URL}/v1/me/player",
        status=200,
        body=load_fixture(f"playback_{playback_id}.json"),
    )
    response = await authenticated_client.get_playback()
    assert response == snapshot
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player",
        METH_GET,
        headers=HEADERS,
        params=None,
        data=None,
    )


async def test_get_no_playback_state(
    authenticated_client: SpotifyClient,
    responses: aioresponses,
) -> None:
    """Test retrieving no playback state."""
    responses.get(
        f"{SPOTIFY_URL}/v1/me/player",
        status=204,
    )
    response = await authenticated_client.get_playback()
    assert response is None
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player",
        METH_GET,
        headers=HEADERS,
        data=None,
        params=None,
    )


async def test_transfer_playback(
    authenticated_client: SpotifyClient,
    responses: aioresponses,
) -> None:
    """Test transferring playback."""
    responses.put(
        f"{SPOTIFY_URL}/v1/me/player",
        status=204,
    )
    await authenticated_client.transfer_playback("test")
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player",
        METH_PUT,
        headers=HEADERS,
        data={"device_ids": ["test"]},
        params=None,
    )


async def test_get_devices(
    snapshot: SnapshotAssertion,
    authenticated_client: SpotifyClient,
    responses: aioresponses,
) -> None:
    """Test retrieving devices."""
    responses.get(
        f"{SPOTIFY_URL}/v1/me/player/devices",
        status=200,
        body=load_fixture("devices.json"),
    )
    devices = await authenticated_client.get_devices()
    assert devices == snapshot
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/devices",
        headers=HEADERS,
        data=None,
        params=None,
    )


async def test_get_current_playing(
    responses: aioresponses,
    snapshot: SnapshotAssertion,
    authenticated_client: SpotifyClient,
) -> None:
    """Test retrieving current playing."""
    responses.get(
        f"{SPOTIFY_URL}/v1/me/player/currently-playing",
        status=200,
        body=load_fixture("current_playing_track.json"),
    )
    response = await authenticated_client.get_current_playing()
    assert response == snapshot
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/currently-playing",
        headers=HEADERS,
        data=None,
        params=None,
    )


async def test_get_no_current_playing_state(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
) -> None:
    """Test retrieving no current playing state."""
    responses.get(
        f"{SPOTIFY_URL}/v1/me/player/currently-playing",
        status=204,
    )
    response = await authenticated_client.get_current_playing()
    assert response is None
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/currently-playing",
        headers=HEADERS,
        params=None,
        data=None,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params", "expected_data"),
    [
        ({}, {}, {"position_ms": 0}),
        ({"device_id": "123qwe"}, {"device_id": "123qwe"}, {"position_ms": 0}),
        (
            {"context_uri": "spotify:artist:6cmp7ut7okJAgJOSaMAVf3"},
            {},
            {"position_ms": 0, "context_uri": "spotify:artist:6cmp7ut7okJAgJOSaMAVf3"},
        ),
        (
            {
                "uris": [
                    "spotify:track:1FyXbzOlq3dkxaB6iRsETv",
                    "spotify:track:4e9hUiLsN4mx61ARosFi7p",
                ],
                "uri_offset": "spotify:track:4e9hUiLsN4mx61ARosFi7p",
            },
            {},
            {
                "position_ms": 0,
                "uris": [
                    "spotify:track:1FyXbzOlq3dkxaB6iRsETv",
                    "spotify:track:4e9hUiLsN4mx61ARosFi7p",
                ],
                "offset": {"uri": "spotify:track:4e9hUiLsN4mx61ARosFi7p"},
            },
        ),
        (
            {"uris": ["spotify:artist:6cmp7ut7okJAgJOSaMAVf3"], "position_offset": 5},
            {},
            {
                "position_ms": 0,
                "uris": ["spotify:artist:6cmp7ut7okJAgJOSaMAVf3"],
                "offset": {"position": 5},
            },
        ),
        ({"position": 5000}, {}, {"position_ms": 5000}),
    ],
)
async def test_resume_playback(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
    expected_data: dict[str, Any],
) -> None:
    """Test resuming playback."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/play",
        query=expected_params,
    )
    responses.put(
        url,
        status=204,
    )
    await authenticated_client.start_playback(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/play",
        METH_PUT,
        headers=HEADERS,
        params=expected_params,
        data=expected_data,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params"),
    [
        ({}, {}),
        ({"device_id": "123qwe"}, {"device_id": "123qwe"}),
    ],
)
async def test_pause_playback(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
) -> None:
    """Test pausing playback."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/pause",
        query=expected_params,
    )
    responses.put(
        url,
        status=204,
    )
    await authenticated_client.pause_playback(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/pause",
        METH_PUT,
        headers=HEADERS,
        params=expected_params,
        data=None,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params"),
    [
        ({}, {}),
        ({"device_id": "123qwe"}, {"device_id": "123qwe"}),
    ],
)
async def test_next_track(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
) -> None:
    """Test next track."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/next",
        query=expected_params,
    )
    responses.post(
        url,
        status=204,
    )
    await authenticated_client.next_track(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/next",
        METH_POST,
        headers=HEADERS,
        params=expected_params,
        data=None,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params"),
    [
        ({}, {}),
        ({"device_id": "123qwe"}, {"device_id": "123qwe"}),
    ],
)
async def test_previous_track(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
) -> None:
    """Test previous track."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/previous",
        query=expected_params,
    )
    responses.post(
        url,
        status=204,
    )
    await authenticated_client.previous_track(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/previous",
        METH_POST,
        headers=HEADERS,
        params=expected_params,
        data=None,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params"),
    [
        ({"position": 5000}, {"position_ms": 5000}),
        (
            {"position": 5000, "device_id": "123qwe"},
            {"device_id": "123qwe", "position_ms": 5000},
        ),
    ],
)
async def test_seek_track(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
) -> None:
    """Test seeking track."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/seek",
        query=expected_params,
    )
    responses.put(
        url,
        status=204,
    )
    await authenticated_client.seek_track(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/seek",
        METH_PUT,
        headers=HEADERS,
        params=expected_params,
        data=None,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params"),
    [
        ({"state": RepeatMode.OFF}, {"state": "off"}),
        ({"state": RepeatMode.TRACK}, {"state": "track"}),
        ({"state": RepeatMode.CONTEXT}, {"state": "context"}),
        (
            {"state": RepeatMode.CONTEXT, "device_id": "123qwe"},
            {"state": "context", "device_id": "123qwe"},
        ),
    ],
)
async def test_set_repeat(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
) -> None:
    """Test setting repeat."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/repeat",
        query=expected_params,
    )
    responses.put(
        url,
        status=204,
    )
    await authenticated_client.set_repeat(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/repeat",
        METH_PUT,
        headers=HEADERS,
        params=expected_params,
        data=None,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params"),
    [
        ({"volume": 50}, {"volume_percent": 50}),
        (
            {"volume": 50, "device_id": "123qwe"},
            {"volume_percent": 50, "device_id": "123qwe"},
        ),
    ],
)
async def test_set_volume(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
) -> None:
    """Test setting volume."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/volume",
        query=expected_params,
    )
    responses.put(
        url,
        status=204,
    )
    await authenticated_client.set_volume(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/volume",
        METH_PUT,
        headers=HEADERS,
        params=expected_params,
        data=None,
    )


@pytest.mark.parametrize(
    ("arguments", "expected_params"),
    [
        ({"state": True}, {"state": "true"}),
        (
            {"state": False, "device_id": "123qwe"},
            {"state": "false", "device_id": "123qwe"},
        ),
    ],
)
async def test_set_shuffle(
    responses: aioresponses,
    authenticated_client: SpotifyClient,
    arguments: dict[str, Any],
    expected_params: dict[str, Any],
) -> None:
    """Test setting shuffle."""
    url = URL.build(
        scheme="https",
        host="api.spotify.com",
        port=443,
        path="/v1/me/player/shuffle",
        query=expected_params,
    )
    responses.put(
        url,
        status=204,
    )
    await authenticated_client.set_shuffle(**arguments)
    responses.assert_called_once_with(
        f"{SPOTIFY_URL}/v1/me/player/shuffle",
        METH_PUT,
        headers=HEADERS,
        params=expected_params,
        data=None,
    )
