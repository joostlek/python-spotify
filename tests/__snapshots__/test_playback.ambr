# serializer version: 1
# name: test_get_current_playing
  dict({
    'context': dict({
      'context_type': <ContextType.ARTIST: 'artist'>,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/artist/6mS5GeFyhea6w9OKo8PO3p',
      }),
      'href': 'https://api.spotify.com/v1/artists/6mS5GeFyhea6w9OKo8PO3p',
      'uri': 'spotify:artist:6mS5GeFyhea6w9OKo8PO3p',
    }),
    'currently_playing_type': 'track',
    'is_playing': True,
    'item': dict({
      'album': dict({
        'album_id': '1LorgdkFY2zgE6NbDdY19k',
        'album_type': <AlbumType.SINGLE: 'single'>,
        'artists': list([
          dict({
            'artist_id': '6mS5GeFyhea6w9OKo8PO3p',
            'name': 'Goldband',
            'uri': 'spotify:artist:6mS5GeFyhea6w9OKo8PO3p',
          }),
        ]),
        'images': list([
          dict({
            'height': 640,
            'url': 'https://i.scdn.co/image/ab67616d0000b2731de89c5340be36804e01fdf2',
            'width': 640,
          }),
          dict({
            'height': 300,
            'url': 'https://i.scdn.co/image/ab67616d00001e021de89c5340be36804e01fdf2',
            'width': 300,
          }),
          dict({
            'height': 64,
            'url': 'https://i.scdn.co/image/ab67616d000048511de89c5340be36804e01fdf2',
            'width': 64,
          }),
        ]),
        'name': 'Dun Smeren Geld Verdienen',
        'release_date': '2019-05-13',
        'release_date_precision': <ReleaseDatePrecision.DAY: 'day'>,
        'total_tracks': 6,
        'uri': 'spotify:album:1LorgdkFY2zgE6NbDdY19k',
      }),
      'artists': list([
        dict({
          'artist_id': '6mS5GeFyhea6w9OKo8PO3p',
          'name': 'Goldband',
          'uri': 'spotify:artist:6mS5GeFyhea6w9OKo8PO3p',
        }),
      ]),
      'disc_number': 1,
      'duration_ms': 245581,
      'explicit': True,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/track/4sIkwgf9bU7rQPvsA37rVQ',
      }),
      'href': 'https://api.spotify.com/v1/tracks/4sIkwgf9bU7rQPvsA37rVQ',
      'is_local': False,
      'name': 'Witte Was',
      'track_id': '4sIkwgf9bU7rQPvsA37rVQ',
      'track_number': 6,
      'uri': 'spotify:track:4sIkwgf9bU7rQPvsA37rVQ',
    }),
    'progress_ms': 134749,
  })
# ---
# name: test_get_devices
  list([
    dict({
      'device_id': '21dac6b0e0a1f181870fdc9749b2656466557687',
      'device_type': <DeviceType.COMPUTER: 'Computer'>,
      'is_active': False,
      'is_private_session': False,
      'is_restricted': False,
      'name': 'DESKTOP-BKC5SIK',
      'supports_volume': True,
      'volume_percent': 69,
    }),
  ])
# ---
# name: test_get_playback_state[1]
  dict({
    'context': dict({
      'context_type': <ContextType.COLLECTION: 'collection'>,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/collection/tracks',
      }),
      'href': 'https://api.spotify.com/v1/me/tracks',
      'uri': 'spotify:user:1112264649:collection',
    }),
    'currently_playing_type': 'track',
    'device': dict({
      'device_id': '21dac6b0e0a1f181870fdc9749b2656466557687',
      'device_type': <DeviceType.COMPUTER: 'Computer'>,
      'is_active': True,
      'is_private_session': False,
      'is_restricted': False,
      'name': 'DESKTOP-BKC5SIK',
      'supports_volume': True,
      'volume_percent': 69,
    }),
    'is_playing': True,
    'item': dict({
      'album': dict({
        'album_id': '1iasbpTobDPa5BmsK0Rz1f',
        'album_type': <AlbumType.SINGLE: 'single'>,
        'artists': list([
          dict({
            'artist_id': '6cmp7ut7okJAgJOSaMAVf3',
            'name': 'Machinae Supremacy',
            'uri': 'spotify:artist:6cmp7ut7okJAgJOSaMAVf3',
          }),
        ]),
        'images': list([
          dict({
            'height': 640,
            'url': 'https://i.scdn.co/image/ab67616d0000b273326db007f167499d23171722',
            'width': 640,
          }),
          dict({
            'height': 300,
            'url': 'https://i.scdn.co/image/ab67616d00001e02326db007f167499d23171722',
            'width': 300,
          }),
          dict({
            'height': 64,
            'url': 'https://i.scdn.co/image/ab67616d00004851326db007f167499d23171722',
            'width': 64,
          }),
        ]),
        'name': 'WARRIORS, Pt. 1 (Final Stage)',
        'release_date': '2023-10-20',
        'release_date_precision': <ReleaseDatePrecision.DAY: 'day'>,
        'total_tracks': 1,
        'uri': 'spotify:album:1iasbpTobDPa5BmsK0Rz1f',
      }),
      'artists': list([
        dict({
          'artist_id': '6cmp7ut7okJAgJOSaMAVf3',
          'name': 'Machinae Supremacy',
          'uri': 'spotify:artist:6cmp7ut7okJAgJOSaMAVf3',
        }),
      ]),
      'disc_number': 1,
      'duration_ms': 268525,
      'explicit': False,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/track/1FyXbzOlq3dkxaB6iRsETv',
      }),
      'href': 'https://api.spotify.com/v1/tracks/1FyXbzOlq3dkxaB6iRsETv',
      'is_local': False,
      'name': 'WARRIORS, Pt. 1 (Final Stage)',
      'track_id': '1FyXbzOlq3dkxaB6iRsETv',
      'track_number': 1,
      'uri': 'spotify:track:1FyXbzOlq3dkxaB6iRsETv',
    }),
    'progress_ms': 225564,
    'repeat_mode': <RepeatMode.OFF: 'off'>,
    'shuffle': False,
  })
# ---
# name: test_get_playback_state[2]
  dict({
    'context': dict({
      'context_type': <ContextType.PLAYLIST: 'playlist'>,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/playlist/2r35vbe6hHl6yDSMfjKgmm',
      }),
      'href': 'https://api.spotify.com/v1/playlists/2r35vbe6hHl6yDSMfjKgmm',
      'uri': 'spotify:user:rushofficial:playlist:2r35vbe6hHl6yDSMfjKgmm',
    }),
    'currently_playing_type': 'track',
    'device': dict({
      'device_id': 'a19f7a03a25aff3e43f457a328a8ba67a8c44789',
      'device_type': <DeviceType.SPEAKER: 'Speaker'>,
      'is_active': True,
      'is_private_session': False,
      'is_restricted': False,
      'name': 'Master Bathroom Speaker',
      'supports_volume': True,
      'volume_percent': 25,
    }),
    'is_playing': True,
    'item': dict({
      'album': dict({
        'album_id': '3nUNxSh2szhmN7iifAKv5i',
        'album_type': <AlbumType.ALBUM: 'album'>,
        'artists': list([
          dict({
            'artist_id': '2Hkut4rAAyrQxRdof7FVJq',
            'name': 'Rush',
            'uri': 'spotify:artist:2Hkut4rAAyrQxRdof7FVJq',
          }),
        ]),
        'images': list([
          dict({
            'height': 640,
            'url': 'https://i.scdn.co/image/ab67616d0000b27306c0d7ebcabad0c39b566983',
            'width': 640,
          }),
          dict({
            'height': 300,
            'url': 'https://i.scdn.co/image/ab67616d00001e0206c0d7ebcabad0c39b566983',
            'width': 300,
          }),
          dict({
            'height': 64,
            'url': 'https://i.scdn.co/image/ab67616d0000485106c0d7ebcabad0c39b566983',
            'width': 64,
          }),
        ]),
        'name': 'Permanent Waves',
        'release_date': '1980-01-01',
        'release_date_precision': <ReleaseDatePrecision.DAY: 'day'>,
        'total_tracks': 6,
        'uri': 'spotify:album:3nUNxSh2szhmN7iifAKv5i',
      }),
      'artists': list([
        dict({
          'artist_id': '2Hkut4rAAyrQxRdof7FVJq',
          'name': 'Rush',
          'uri': 'spotify:artist:2Hkut4rAAyrQxRdof7FVJq',
        }),
      ]),
      'disc_number': 1,
      'duration_ms': 296466,
      'explicit': False,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/track/4e9hUiLsN4mx61ARosFi7p',
      }),
      'href': 'https://api.spotify.com/v1/tracks/4e9hUiLsN4mx61ARosFi7p',
      'is_local': False,
      'name': 'The Spirit Of Radio',
      'track_id': '4e9hUiLsN4mx61ARosFi7p',
      'track_number': 1,
      'uri': 'spotify:track:4e9hUiLsN4mx61ARosFi7p',
    }),
    'progress_ms': 249367,
    'repeat_mode': <RepeatMode.OFF: 'off'>,
    'shuffle': False,
  })
# ---
# name: test_get_playback_state[3]
  dict({
    'context': dict({
      'context_type': <ContextType.COLLECTION: 'collection'>,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/collection/tracks',
      }),
      'href': 'https://api.spotify.com/v1/me/tracks',
      'uri': 'spotify:user:1112264649:collection',
    }),
    'currently_playing_type': 'track',
    'device': dict({
      'device_id': None,
      'device_type': <DeviceType.SPEAKER: 'Speaker'>,
      'is_active': True,
      'is_private_session': False,
      'is_restricted': True,
      'name': 'Sonos Roam SL',
      'supports_volume': True,
      'volume_percent': 34,
    }),
    'is_playing': True,
    'item': dict({
      'album': dict({
        'album_id': '3qCsGHHWG6t9izvmsE47jr',
        'album_type': <AlbumType.SINGLE: 'single'>,
        'artists': list([
          dict({
            'artist_id': '6s5ubAp65wXoTZefE01RNR',
            'name': 'Joost',
            'uri': 'spotify:artist:6s5ubAp65wXoTZefE01RNR',
          }),
        ]),
        'images': list([
          dict({
            'height': 640,
            'url': 'https://i.scdn.co/image/ab67616d0000b273ae0956d669aca836c4fb5a0c',
            'width': 640,
          }),
          dict({
            'height': 300,
            'url': 'https://i.scdn.co/image/ab67616d00001e02ae0956d669aca836c4fb5a0c',
            'width': 300,
          }),
          dict({
            'height': 64,
            'url': 'https://i.scdn.co/image/ab67616d00004851ae0956d669aca836c4fb5a0c',
            'width': 64,
          }),
        ]),
        'name': 'Ome Robert',
        'release_date': '2018-04-26',
        'release_date_precision': <ReleaseDatePrecision.DAY: 'day'>,
        'total_tracks': 1,
        'uri': 'spotify:album:3qCsGHHWG6t9izvmsE47jr',
      }),
      'artists': list([
        dict({
          'artist_id': '6s5ubAp65wXoTZefE01RNR',
          'name': 'Joost',
          'uri': 'spotify:artist:6s5ubAp65wXoTZefE01RNR',
        }),
      ]),
      'disc_number': 1,
      'duration_ms': 152500,
      'explicit': True,
      'external_urls': dict({
        'spotify': 'https://open.spotify.com/track/3TE49HXyoNczZk2IBhP87z',
      }),
      'href': 'https://api.spotify.com/v1/tracks/3TE49HXyoNczZk2IBhP87z',
      'is_local': False,
      'name': 'Ome Robert',
      'track_id': '3TE49HXyoNczZk2IBhP87z',
      'track_number': 1,
      'uri': 'spotify:track:3TE49HXyoNczZk2IBhP87z',
    }),
    'progress_ms': 7919,
    'repeat_mode': <RepeatMode.OFF: 'off'>,
    'shuffle': False,
  })
# ---
